File: backend/inmobiliaria-api/inmobiliaria-api/src/main/java/com/inmobiliaria/inmobiliaria_api/service/FacturaService.java
package com.inmobiliaria.inmobiliaria_api.service;

import com.inmobiliaria.inmobiliaria_api.dto.request.FacturaRequest;
import com.inmobiliaria.inmobiliaria_api.dto.response.FacturaResponse;
import com.inmobiliaria.inmobiliaria_api.dto.response.PageResponse;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface FacturaService {

    FacturaResponse guardar(FacturaRequest request);

    PageResponse<FacturaResponse> listar(Pageable pageable);

    FacturaResponse buscarPorId(Long idFactura);

    List<FacturaResponse> listarPorContrato(Long idContrato);

    FacturaResponse actualizar(Long idFactura, FacturaRequest request);

    void eliminar(Long idFactura);
}

File: frontend/inmobiliaria-web/src/app/features/facturas/services/factura.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { Factura, FacturaRequest } from '../models/factura.model';

export interface PageResponse<T> {
  contenido: T[];
  paginaActual: number;
  tamanoPagina: number;
  totalRegistros: number;
  totalPaginas: number;
  primera: boolean;
  ultima: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class FacturaService {

  private readonly http = inject(HttpClient);

  private readonly apiUrl = 'http://localhost:8080/api/facturas';

  listar(
    page = 0,
    size = 10,
    sortBy = 'idFactura',
    direction = 'asc'
  ): Observable<PageResponse<Factura>> {

    const params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sortBy', sortBy)
      .set('direction', direction);

    return this.http.get<PageResponse<Factura>>(
      this.apiUrl,
      { params }
    );
  }

  buscarPorId(idFactura: number): Observable<Factura> {
    return this.http.get<Factura>(
      `${this.apiUrl}/${idFactura}`
    );
  }

  listarPorContrato(idContrato: number): Observable<Factura[]> {
    return this.http.get<Factura[]>(
      `${this.apiUrl}/contrato/${idContrato}`
    );
  }

  crear(request: FacturaRequest): Observable<Factura> {
    return this.http.post<Factura>(
      this.apiUrl,
      request
    );
  }

  actualizar(
    idFactura: number,
    request: FacturaRequest
  ): Observable<Factura> {

    return this.http.put<Factura>(
      `${this.apiUrl}/${idFactura}`,
      request
    );
  }

  eliminar(idFactura: number): Observable<void> {
    return this.http.delete<void>(
      `${this.apiUrl}/${idFactura}`
    );
  }
}
